### ISSUE TITLE
STATEMENT error when mixing rule types (Byte Match Rules)

### ISSUE THREAD

**Issue by voycey** [umotif-public/terraform-aws-waf-webaclv2#23]
**What is the current behavior?**

When trying to add a byte match rule I am getting errors on apply: 

```
│ Error: Error updating WAFv2 WebACL: WAFInvalidParameterException: Error reason: You have used none or multiple values for a field that requires exactly one value., field: STATEMENT, parameter: Statement
│ {
│   RespMetadata: {
│     StatusCode: 400,
│     RequestID: "c597c3a8-fe66-4166-bad8-ecd26d218587"
│   },
│   Field: "STATEMENT",
│   Message_: "Error reason: You have used none or multiple values for a field that requires exactly one value., field: STATEMENT, parameter: Statement",
│   Parameter: "Statement",
│   Reason: "You have used none or multiple values for a field that requires exactly one value."
│ }
│ 
│   with module.wafv2-regional.module.wafreg.aws_wafv2_web_acl.main[0],
│   on .terraform/modules/wafv2-regional.wafreg/main.tf line 4, in resource "aws_wafv2_web_acl" "main":
│    4: resource "aws_wafv2_web_acl" "main" {
│ 
```
If I remove the byte match rule everything functions correctly.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**
```
#####
# IP set resources
#####
resource "aws_wafv2_ip_set" "blacklisted_ips" {
  name = "${var.stage}-blacklisted-ips"

  scope              = "REGIONAL"
  ip_address_version = "IPV4"

  addresses = var.blacklisted_ips
}

resource "aws_wafv2_ip_set" "whitelisted_ips" {
  name = "${var.stage}-whitelisted-ips"

  scope              = "REGIONAL"
  ip_address_version = "IPV4"

  addresses = var.whitelisted_ips
}

module "wafreg" {
  source  = "umotif-public/waf-webaclv2/aws"
  version = "~> 1.5.1"

  name_prefix = "${var.stage}-wafv2-regional"
  alb_arn     = var.svc_lb_arn

  scope = "REGIONAL"

  create_alb_association = true

  allow_default_action = true # set to allow if not specified

  visibility_config = {
    metric_name = "${var.stage}-waf-setup-wafv2-main-metrics"
  }

  rules = [
    {
      name     = "AWSManagedRulesCommonRuleSet-${var.stage}-rule-1"
      priority = "1"

      override_action = "none" # set to none if not specified

      visibility_config = {
        metric_name = "AWSManagedRulesCommonRuleSet-${var.stage}-metric"
      }

      managed_rule_group_statement = {
        name        = "AWSManagedRulesCommonRuleSet"
        vendor_name = "AWS"
        excluded_rule = [
          "SizeRestrictions_BODY",
        ]
      }
    },
    {
      name     = "AWSManagedRulesKnownBadInputsRuleSet-${var.stage}-rule-2"
      priority = "2"

      override_action = "count"

      visibility_config = {
        metric_name = "AWSManagedRulesKnownBadInputsRuleSet-${var.stage}-metric"
      }

      managed_rule_group_statement = {
        name        = "AWSManagedRulesKnownBadInputsRuleSet"
        vendor_name = "AWS"
      }
    },
    {
      name     = "AWSManagedRulesPHPRuleSet-${var.stage}-rule-3"
      priority = "3"

      visibility_config = {
        cloudwatch_metrics_enabled = false
        metric_name                = "AWSManagedRulesPHPRuleSet-${var.stage}-metric"
        sampled_requests_enabled   = false
      }

      managed_rule_group_statement = {
        name        = "AWSManagedRulesPHPRuleSet"
        vendor_name = "AWS"
      }
    },
    {
      name     = "block-specific-uri-path"
      priority = "4"
      action   = "block"

      byte_match_statement = {
        field_to_match = {
          uri_path = "{}"
        }
        positional_constraint = "STARTS_WITH"
        search_string         = "/path/to/match"
        priority              = 0
        type                  = "NONE" # The text transformation type
      }

      visibility_config = {
        cloudwatch_metrics_enabled = false
        sampled_requests_enabled   = false
      }
    }
  ]

  ip_set_rules = [
    {
      name     = "${var.stage}-truebiz-wafv2-ip-set-rule"
      priority = "5"
      action   = "allow" # if not set, action defaults to allow
      ip_set_reference_statement = {
        arn = aws_wafv2_ip_set.whitelisted_ips.arn
      }

      visibility_config = {
        cloudwatch_metrics_enabled = false
        sampled_requests_enabled   = false
      }
    },
    {
      name     = "${var.stage}-truebiz-wafv2-blacklisted-ip-set"
      priority = "6"
      action   = "block"
      ip_set_reference_statement = {
        arn = aws_wafv2_ip_set.blacklisted_ips.arn
      }

      visibility_config = {
        cloudwatch_metrics_enabled = false
        metric_name                = "${var.stage}-truebiz-ip-set-metrics"
        sampled_requests_enabled   = false
      }
    }
  ]

  tags = {
    "Name" = "${var.stage}-wafv2-regional"
  }
}
```

**What is the expected behavior?**
The rules apply

**Software versions?**

Terraform v0.15.4
on darwin_amd64
+ provider registry.terraform.io/hashicorp/aws v3.42.0
+ provider registry.terraform.io/hashicorp/random v3.1.0

Same behaviour on -> 1.5.1 and -> 2.0

**Comment by marcincuber** [comment]
@voycey Apologies for this error. This is because we have changes in Readme file that are actually part of `3.0.0`. That particular release will fix the issue you are reporting. In fact, you will be able to define all rules inside `rules` list. 

Release `3.0.0` will be available once #22 is resolved.

For the time being, could you try running your configuration against master branch of this repo?

Thanks

**Comment by voycey** [comment]
Thanks, I have tried running against master and still getting the same
errors unfortunately, had to change the syntax a little bit to remove the
seperate IP rules but still same error.

If I comment out the bytematch everything works fine that currently exists
there!

**Comment by seanpascual** [comment]
Hi @voycey thanks for adding the single_header option for byte_match statements. I have now merged your change in. :)

Once the tests in #22 are resolved, I will release this as 3.0.0.
